src/app/Rules/CreateRule.tsx

```diff
@@ -274,38 +277,55 @@ const CreateRuleForm: React.FC<CreateRuleFormProps> = ({ ...props }) => {
         .subscribe((templates) => {
           setTemplates(templates);
           setTemplate((old) => {
-            const matched = templates.find((t) => t.name === old.name && t.type === t.type);
+            const matched = templates.find((t) => t.name === old.name && t.type === old.type);
             return matched ? { name: matched.name, type: matched.type } : {};
           });
         })
     );
-  }, [addSubscription, context.api, matchedTargets]);
+  }, [addSubscription, context.api]);
 
   React.useEffect(() => {
-    addSubscription(context.targets.targets().subscribe(setTargets));
-  }, [addSubscription, context.targets, setTargets]);
-
-  React.useEffect(() => {
-    // Set validations
-    let validation: ValidatedOptions = ValidatedOptions.default;
-    let matches: Target[] = [];
-    if (matchExpression !== '' && targets.length > 0) {
-      try {
-        matches = targets.filter((t) => {
-          const res = evaluateTargetWithExpr(t, matchExpression);
-          if (typeof res === 'boolean') {
-            return res;
-          }
-          throw new Error('The expression matching failed.');
-        });
-        validation = matches.length ? ValidatedOptions.success : ValidatedOptions.warning;
-      } catch (err) {
-        validation = ValidatedOptions.error;
-      }
-    }
-    setMatchExpressionValid(validation);
-    matchedTargets.next(matches);
-  }, [matchExpression, targets, matchedTargets, setMatchExpressionValid]);
+    const matchedTargets = matchedTargetsRef.current;
+    addSubscription(
+      combineLatest([
+        matchExprService.searchExpression(),
+        context.targets.targets().pipe(tap((ts) => setSampleTarget(ts[0]))),
+      ])
+        .pipe(
+          switchMap(([input, targets]) => {
+            setEvaluating(true);
```

Comment:
Can this state get set as soon as the text input element sees a keystroke, rather than getting set when the API request is about to be fired? ie. it shouldn't be subject to the debounce time

Reply:
Oh yeh it should be possible.

Just that `matchExprService.searchExpression()` need refactoring to remove internal default debouceTime and each consumer needs to specify there own.

Reply:
ah mvnind found a better way :))

Reply:
Updated :))